File: deck/base/code/native/kotlin/runtime/x25519.kt
import java.security.KeyFactory
import java.security.KeyPairGenerator
import java.security.interfaces.XECPrivateKey
import java.security.spec.PKCS8EncodedKeySpec
import java.security.spec.X509EncodedKeySpec
import javax.crypto.KeyAgreement

// X25519 ECDH over java.security (built into the JDK since 11, no extra dependency). The raw 32-byte private key is
// read straight off the key (XECPrivateKey.scalar); the 32-byte public key is the tail of its X.509 encoding. To
// rebuild a key from raw bytes the fixed X25519 DER prefixes are prepended (PKCS8 for private, X.509 for public).
object x25519 {
    private const val PKCS8_PREFIX = "302e020100300506032b656e04220420"
    private const val SPKI_PREFIX = "302a300506032b656e032100"
    private fun fromHex(hex: String): ByteArray = ByteArray(hex.length / 2) { hex.substring(it * 2, it * 2 + 2).toInt(16).toByte() }
    private fun toHex(bytes: ByteArray): String = bytes.joinToString("") { "%02x".format(it) }
    fun makeKeyPair(): String {
        val pair = KeyPairGenerator.getInstance("X25519").generateKeyPair()
        val scalar = (pair.private as XECPrivateKey).scalar.get()
        val encoded = pair.public.encoded
        val public = encoded.copyOfRange(encoded.size - 32, encoded.size)
        return toHex(scalar) + toHex(public)
    }
    fun sharedSecret(privateHex: String, publicHex: String): String {
        val privateKey = KeyFactory.getInstance("X25519").generatePrivate(PKCS8EncodedKeySpec(fromHex(PKCS8_PREFIX + privateHex)))
        val publicKey = KeyFactory.getInstance("X25519").generatePublic(X509EncodedKeySpec(fromHex(SPKI_PREFIX + publicHex)))
        val agreement = KeyAgreement.getInstance("X25519")
        agreement.init(privateKey)
        agreement.doPhase(publicKey, true)
        return toHex(agreement.generateSecret())
    }
}

File: deck/base/code/native/swift/runtime/x25519.swift
import Foundation
import CryptoKit

// X25519 ECDH over CryptoKit. The private key and public key are raw 32-byte representations; the shared secret is the
// raw 32-byte X25519 output. make-key-pair returns the private key and public key concatenated as hex.
enum x25519 {
    private static func fromHex(_ hex: String) -> Data {
        var data = Data(capacity: hex.count / 2)
        var index = hex.startIndex
        while index < hex.endIndex {
            let next = hex.index(index, offsetBy: 2)
            data.append(UInt8(hex[index..<next], radix: 16)!)
            index = next
        }
        return data
    }
    private static func toHex(_ data: Data) -> String { data.map { String(format: "%02x", $0) }.joined() }
    static func makeKeyPair() -> String {
        let secret = Curve25519.KeyAgreement.PrivateKey()
        return toHex(secret.rawRepresentation) + toHex(secret.publicKey.rawRepresentation)
    }
    static func sharedSecret(_ privateHex: String, _ publicHex: String) -> String {
        let secret = try! Curve25519.KeyAgreement.PrivateKey(rawRepresentation: fromHex(privateHex))
        let publicKey = try! Curve25519.KeyAgreement.PublicKey(rawRepresentation: fromHex(publicHex))
        let shared = try! secret.sharedSecretFromKeyAgreement(with: publicKey)
        return shared.withUnsafeBytes { toHex(Data($0)) }
    }
}
